src/React/Hook.purs

```diff
@@ -0,0 +1,310 @@
+import Effect.Uncurried (EffectFn1, EffectFn2, runEffectFn1, runEffectFn2)
+
+import Unsafe.Coerce (unsafeCoerce)
+
+useState
+  :: forall a
+   . a
+  -> Hook (Tuple a (SetState a))
+useState = runFn2 useState_ Tuple
+
+useStateLazy
+  :: forall a
+   . (Unit -> a)
+  -> Hook (Tuple a (SetState a))
+useStateLazy = runFn2 useState_ Tuple
+
+setState
+  :: forall a
+   . SetState a
+  -> a
+  -> Effect Unit
+setState k = runEffectFn1 k'
+  where
+  k' :: EffectFn1 a Unit
+  k' = unsafeCoerce k
+
+modifyState
+  :: forall a
+   . SetState a
+  -> (a -> a)
+  -> Effect Unit
+modifyState k = runEffectFn1 k'
+  where
+  k' :: EffectFn1 (a -> a) Unit
+  k' = unsafeCoerce k
+
+foreign import data SetState :: Type -> Type
+
+foreign import useState_
+  :: forall a b
+   . Fn2 (b -> SetState b -> Tuple b (SetState b))
+         a
+         (Hook (Tuple b (SetState b)))
+
+useEffect
+  :: forall a
+   . Effect (Effect a)
+  -> Maybe (Array EffectInput)
+  -> Hook Unit
+useEffect k = runFn2 useEffect_ k <<< Nullable.toNullable
+
+effectInput :: forall a. a -> EffectInput
+effectInput = unsafeCoerce
+
+foreign import data EffectInput :: Type
+
+foreign import useEffect_
+  :: forall a
+   . Fn2 (Effect (Effect a))
+         (Nullable (Array EffectInput))
+         (Hook Unit)
+
+useContext :: forall a. Context a -> Hook a
+useContext = runFn1 useContext_
+
+foreign import data Context :: Type -> Type
+
+foreign import useContext_
+  :: forall a
+   . Fn1 (Context a)
+         (Hook a)
+
+useReducer
+  :: forall a b
+   . (a -> b -> a)
+  -> a
+  -> Hook (Tuple a (Dispatch a b))
+useReducer = runFn3 useReducer_ Tuple <<< mkFn2
+
+useReducerLazy
+  :: forall a b
+   . (a -> b -> a)
+  -> a
+  -> b
+  -> Hook (Tuple a (Dispatch a b))
+useReducerLazy = runFn4 useReducerLazy_ Tuple <<< mkFn2
+
+dispatch
+  :: forall a b
+   . Dispatch a b
+  -> a
+  -> Effect Unit
+dispatch k = runEffectFn1 k'
+  where
+  k' :: EffectFn1 a Unit
+  k' = unsafeCoerce k
+
+foreign import data Dispatch :: Type -> Type -> Type
+
+foreign import useReducer_
+  :: forall a b
+   . Fn3 (a -> Dispatch a b -> Tuple a (Dispatch a b))
+         (Fn2 a b a)
+         a
+         (Hook (Tuple a (Dispatch a b)))
+
+foreign import useReducerLazy_
+  :: forall a b
+   . Fn4 (a -> Dispatch a b -> Tuple a (Dispatch a b))
+         (Fn2 a b a)
+         a
+         b
+         (Hook (Tuple a (Dispatch a b)))
+
+useCallback
+  :: forall a b
+   . (a -> b)
+  -> Maybe (Array CallbackInput)
+  -> Hook (a -> b)
+useCallback k = runFn2 useCallback_ k <<< Nullable.toNullable
+
+callbackInput :: forall a. a -> CallbackInput
+callbackInput = unsafeCoerce
+
+foreign import data CallbackInput :: Type
+
+foreign import useCallback_
+  :: forall a b
+   . Fn2 (a -> b)
+         (Nullable (Array CallbackInput))
+         (Hook (a -> b))
+
+useMemo
+  :: forall a b
+   . (Unit -> a -> b)
+  -> Maybe (Array MemoInput)
+  -> Hook (a -> b)
+useMemo k = runFn2 useMemo_ k <<< Nullable.toNullable
```

Comment:
I don't think `useMemo` necessarily computes a function. It is supposed to memoize an arbitrary value (`useCallback` is just a shortcut for memoizing a function).

Reply:
Gotcha. I will review this. I haven't played with it too much yet. Thanks.